### PR TITLE
Update logs, importer, login & logout

### PR DESCRIPTION
**Pull request by juanvallejo**
Picks and rebases changes from https://github.com/openshift/origin/pull/20232

cc @deads2k

### CODE REVIEW

**Comment by deads2k** [comment]
find what is changing completions and make a different pull with just that.  Keep this one without adjusting completions, but squash it down.

File: pkg/oc/cli/cmd/image/mirror/mirror.go [hunk 1]
```diff
@@ -98,47 +102,49 @@ type pushOptions struct {
...
-	flag.BoolVar(&o.SkipMount, "skip-mount", o.SkipMount, "Always push layers instead of cross-mounting them")
-	flag.BoolVar(&o.SkipMultipleScopes, "skip-multiple-scopes", o.SkipMultipleScopes, "Some registries do not support multiple scopes passed to the registry login.")
-	flag.StringVar(&o.FilterByOS, "filter-by-os", o.FilterByOS, "A regular expression to control which images are mirrored. Images will be passed as '<platform>/<architecture>[/<variant>]'.")
-	flag.BoolVar(&o.Force, "force", o.Force, "Attempt to write all layers and manifests even if they exist in the remote repository.")
-	flag.IntVar(&o.MaxRegistry, "max-registry", 4, "Number of concurrent registries to connect to at any one time.")
-	flag.IntVar(&o.MaxPerRegistry, "max-per-registry", 6, "Number of concurrent requests allowed per registry.")
-	flag.StringSliceVar(&o.AttemptS3BucketCopy, "s3-source-bucket", o.AttemptS3BucketCopy, "A list of bucket/path locations on S3 that may contain already uploaded blobs. Add [store] to the end to use the Docker registry path convention.")
-	flag.StringSliceVarP(&o.Filenames, "filename", "f", o.Filenames, "One or more files to read SRC=DST or SRC DST [DST ...] mappings from.")
```
Comment (deads2k) [inline pkg/oc/cli/cmd/image/mirror/mirror.go]: This change looks suspicious. You sure the -f for this command means the same thing?

File: pkg/oc/cli/cmd/importer/appjson.go [hunk 2]
```diff
@@ -56,67 +59,68 @@ var (
...
+	resource.FilenameOptions
 }
 
-// NewCmdAppJSON imports an app.json file (schema described here: https://devcenter.heroku.com/articles/app-json-schema)
-// as a template.
-func NewCmdAppJSON(fullName string, f kcmdutil.Factory, in io.Reader, out, errout io.Writer) *cobra.Command {
-	options := &AppJSONOptions{
+func NewAppJSONOptions(streams genericclioptions.IOStreams) *AppJSONOptions {
+	return &AppJSONOptions{
 		Action: configcmd.BulkAction{
```
Comment (deads2k) [inline pkg/oc/cli/cmd/importer/appjson.go]: Change the bulkAction

File: pkg/oc/cli/cmd/importer/appjson.go [hunk 3]
```diff
@@ -56,67 +59,68 @@ var (
...
-	cmd.Flags().StringVar(&options.BaseImage, "image", options.BaseImage, "An optional image to use as your base Docker build (must have ONBUILD directives)")
-	cmd.Flags().String("generator", options.Generator, "The name of the generator strategy to use - specify this value to for backwards compatibility.")
-	cmd.Flags().StringVar(&options.AsTemplate, "as-template", "", "If set, generate a template with the provided name")
-
-	options.Action.BindForOutput(cmd.Flags())
-	cmd.Flags().String("output-version", "", "The preferred API versions of the output objects")
+	cmd.Flags().StringVar(&o.BaseImage, "image", o.BaseImage, "An optional image to use as your base Docker build (must have ONBUILD directives)")
+	cmd.Flags().StringVar(&o.Generator, "generator", o.Generator, "The name of the generator strategy to use - specify this value to for backwards compatibility.")
+	cmd.Flags().StringVar(&o.AsTemplate, "as-template", o.AsTemplate, "If set, generate a template with the provided name")
+	cmd.Flags().StringVar(&o.OutputVersionStr, "output-version", o.OutputVersionStr, "The preferred API versions of the output objects")
```
Comment (deads2k) [inline pkg/oc/cli/cmd/importer/appjson.go]: followup issue to see about deprecating this

File: pkg/oc/cli/cmd/login/login.go [hunk 4]
```diff
@@ -89,8 +77,8 @@ func NewCmdLogin(fullName string, f kcmdutil.Factory, streams genericclioptions.
 	}
 
 	// Login is the only command that can negotiate a session token against the auth server using basic auth
-	cmds.Flags().StringVarP(&options.Username, "username", "u", "", "Username, will prompt if not provided")
-	cmds.Flags().StringVarP(&options.Password, "password", "p", "", "Password, will prompt if not provided")
+	cmds.Flags().StringVarP(&o.Username, "username", "u", "", "Username, will prompt if not provided")
```
Comment (deads2k) [inline pkg/oc/cli/cmd/login/login.go]: why aren't you using the struct as the default value

**Comment by juanvallejo** [comment]
@deads2k 
> find what is changing completions and make a different pull with just that. Keep this one without adjusting completions, but squash it down.

- `pkg/oc/cli/cmd/importer/appjson.go` adds `o.PrintFlags.AddFlags`, which after rebasing this PR with latest master, adds the --template related flags.

- `pkg/oc/cli/cmd/image/mirror/mirror.go` adds `kcmdutil.AddJsonFilenameFlag(cmd.Flags(), &o.Filenames, usage)` which adds `flags_with_completion+=("--filename")` and `flags_completion+=("__oc_handle_filename_extension_flag json|yaml|yml")`

File: pkg/oc/cli/cmd/image/mirror/mirror.go (hunk 1 above)
Comment (juanvallejo) [reply pkg/oc/cli/cmd/image/mirror/mirror.go]: Not sure what you mean, the `f` shorthand is now added by the call to `kcmdutil.AddJsonFilenameFlag`

File: pkg/oc/cli/cmd/importer/appjson.go (hunk 3 above)
Comment (juanvallejo) [reply pkg/oc/cli/cmd/importer/appjson.go]: Opened issue here: https://github.com/openshift/origin/issues/20277

**Comment by juanvallejo** [comment]
Changes for `image/mirror/mirror.go` reverted.

File: pkg/oc/cli/cmd/importer/appjson.go [hunk 5]
```diff
@@ -56,67 +59,68 @@ var (
...
+func NewAppJSONOptions(streams genericclioptions.IOStreams) *AppJSONOptions {
+	return &AppJSONOptions{
+		bulkAction: configcmd.BulkAction{
+			Out:    streams.Out,
+			ErrOut: streams.ErrOut,
 		},
-		In:        in,
-		Generator: AppJSONV1GeneratorName,
+		IOStreams:  streams,
+		PrintFlags: genericclioptions.NewPrintFlags("").WithTypeSetter(ocscheme.PrintingInternalScheme),
```
Comment (deads2k) [inline pkg/oc/cli/cmd/importer/appjson.go]: "created"?

**Comment by deads2k** [comment]
slip that change into another pull.

/lgtm

File: pkg/oc/cli/cmd/importer/appjson.go (hunk 5 above)
Comment (juanvallejo) [reply pkg/oc/cli/cmd/importer/appjson.go]: Added this change in https://github.com/openshift/origin/pull/20269

**Pull request merged**